[quant] Per Channel quantization APIs

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24936 [quant] per channel quantization support
* **#24935 [quant] Per Channel quantization APIs**
* #24934 [quant] Add missing functions and methods for channelwise quantization

Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
[quant] Per Channel quantization APIs

Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

jerryzh168 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -27,5 +27,38 @@ Tensor empty_affine_quantized_cpu(
       optional_memory_format.value_or(MemoryFormat::Contiguous));
 }
 
+Tensor empty_perchannel_affine_quantized_cpu(
```
nit: per_channel

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -64,6 +64,22 @@ def test_qtensor_quant_dequant(self):
         rqr = qr.dequantize()
         self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
 
+    def test_perchannel_qtensor_creation(self):
```
same here

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

jianyuh commented on test/test_quantized_tensor.py:
```diff
@@ -71,6 +71,22 @@ def test_qtensor_quant_dequant(self):
+        numel = 10
+        ch_axis = 0
+        scales = torch.rand(numel, dtype=torch.double)
+        zero_points = torch.randint(0 , 10, size=(numel,), dtype=torch.long)
```
nit: remove extra space after 0

dskhudia committed:
Update on "[quant] Per Channel quantization APIs"


Adding per channel qtensor creation APIs

Added two tests:

EmptyPerchannelQuantized in aten/src/ATen/test/quantized_test.cpp

test_perchannel_qtensor_creation in test/test_quantized_tensor.py

Differential Revision: [D16696959](https://our.internmc.facebook.com/intern/diff/D16696959/)

jianyuh reviewed (approved):
Looks good!

raghuramank100 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -27,5 +27,38 @@ Tensor empty_affine_quantized_cpu(
+    const TensorOptions& options,
+    c10::optional<c10::MemoryFormat> optional_memory_format) {
+  TORCH_CHECK(
+      options.has_dtype(),
```
Can you check if the dtypes  are qint8/quint8?

raghuramank100 reviewed (approved):
Please add an additional check for dtype
